api/v1alpha1/amaltheasession_children.go

```diff
@@ -0,0 +1,208 @@
+		TerminationMessagePath:   "/dev/termination-log",
+		TerminationMessagePolicy: v1.TerminationMessageReadFile,
+	}
+
+	securityContext := &v1.SecurityContext{
+		RunAsNonRoot: &[]bool{true}[0],
+		RunAsUser:    &[]int64{session.RunAsUser}[0],
+		RunAsGroup:   &[]int64{session.RunAsGroup}[0],
+	}
+
+	if session.RunAsUser == 0 {
+		securityContext.RunAsNonRoot = &[]bool{false}[0]
+	}
+
+	sessionContainer.SecurityContext = securityContext
+
+	containers := []v1.Container{sessionContainer}
+	containers = append(containers, cr.Spec.ExtraContainers...)
+
+	sts := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: v1.PodSpec{
+					Containers:     containers,
+					InitContainers: cr.Spec.ExtraInitContainers,
+					Volumes: []v1.Volume{
+						{
+							Name:         sessionVolumeName,
+							VolumeSource: v1.VolumeSource{PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{ClaimName: pvc.Name}},
+						},
+					},
+				},
+			},
+		},
+	}
+	return sts
+}
+
+// Service returns a AmaltheaSession Service object
+func (cr *AmaltheaSession) Service() v1.Service {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	svc := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name,
+			Namespace: cr.Namespace,
+		},
+		Spec: v1.ServiceSpec{
+			Selector: labels,
+			Ports: []v1.ServicePort{{
+				Protocol:   v1.ProtocolTCP,
+				Name:       servicePortName,
+				Port:       servicePort,
+				TargetPort: intstr.FromInt32(cr.Spec.Session.Port),
+			}},
+		},
+	}
+	return svc
+}
+
+// Ingress returns a AmaltheaSession Ingress object
+func (cr *AmaltheaSession) Ingress() *networkingv1.Ingress {
+	labels := labelsForAmaltheaSession(cr.Name)
+
+	ingress := cr.Spec.Ingress
+
+	ing := &networkingv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        cr.Name,
+			Namespace:   cr.Namespace,
+			Labels:      labels,
+			Annotations: ingress.Annotations,
+		},
+		Spec: networkingv1.IngressSpec{
+			IngressClassName: ingress.IngressClassName,
+			Rules: []networkingv1.IngressRule{{
+				Host: ingress.Host,
+				IngressRuleValue: networkingv1.IngressRuleValue{
+					HTTP: &networkingv1.HTTPIngressRuleValue{
+						Paths: []networkingv1.HTTPIngressPath{{
+							Path: ingress.PathPrefix,
+							PathType: func() *networkingv1.PathType {
+								pt := networkingv1.PathTypePrefix
+								return &pt
+							}(),
+							Backend: networkingv1.IngressBackend{
+								Service: &networkingv1.IngressServiceBackend{
+									Name: cr.Name,
+									Port: networkingv1.ServiceBackendPort{
+										Name: servicePortName,
+									},
+								},
+							},
+						}},
+					},
+				},
+			}},
+			TLS: []networkingv1.IngressTLS{{
```

Comment:
Shouldn't the TLS part depend on the presence of TLSSecretName ?

Reply:
Yes definitely. And to follow the guide on optional vs requried values in CRD I marked that as optional and converted it to a pointer.